proposals/NNNN-reflection.md

```diff
@@ -0,0 +1,704 @@
+both cases, we still have no fundamental way of getting our generic arguments to
+say create a specialized wrapper, get their size, alignment, etc. We could alter
+the language to support these use cases, but the information needed to perform
+these queries are already available and present in Swift binaries.
+
+The unfortunate truth with all of these examples is that the compiler emits the
+data required to achieve what we want in every case. Our current public API just
+doesn't surface any of that vast amount of information that is available to us.
+
+## Proposed solution
+
+Swift introduces a new module, `Reflection`, with a whole new suite of high
+level APIs that make working with reflection easier, more ergonomic, and
+provides developers with much more information than what they had.
+
+We can take each of the motivating examples and see how we can achieve our
+result using these new `Reflection` APIs.
+
+Our registration example can finally iterate a type's fields without even
+having access to an instance of the type by using the new `Type` API:
+
+```swift
+import Reflection
+
+func register(_ type: Any.Type) {
+  for prop in Type(type).storedProperties {
+    if prop.type.swiftType is CustomRegister.Type {
+      cache[type].append(prop)
+    }
+  }
+}
+```
+
+In our enum case name example, we can use a new type called `Case` and use our
+enum instance to initialize a value of this `Case` and grab the name out:
+
+```swift
+import Reflection
+
+enum Food: String, CodingKey {
+  case burger = "Hamburger"
+  case fries = "French Fries"
+  case tots = "Tater tots"
+  case milkshake = "Milkshake"
+}
+
+let favoriteFood = Food.tots
+let caze = Case(from: favoriteFood)!
+
+print(caze.name) // "tots"
+```
+
+Finally, we can use a new type called `PartialType` that lets us do these sort
+of generic erased queries and utilize another new API on `Type` called
+`genericArguments`:
+
+```swift
+import Reflection
+
+func sequenceAlgorithm<T: Sequence>(_ seq: T) {
+  let genericArgs = Type(T.self).genericArguments
+
+  // Void doesn't matter here as you'll see later on
+  if Type(T.self).partial == Type([Void].self).partial {
+    let elementType = genericArgs[0]
+  }
+
+  // This dictionary type doesn't matter here as you'll see later on
+  if Type(T.self).partial == Type([AnyHashable: Any].Keys.self).partial {
+    let keyType = genericArgs[0]
+    let valueType = genericArgs[1]
+  }
+
+  // Do slow thing
+}
+```
+
+## Detailed design
+
+`Reflection` will be a new module included in the Swift toolchain. In order to
+use this module, one must explicitly declare an `import Reflection`.
+
+### `Type`
+
+This is our main entry point into the reflection module. `Type` represents all
+possible concrete types that can occur in Swift. Ranging from structs, enums,
+tuples, functions, existentials, etc. Anything that can be on the right hand
+side of an `as` is represented here in `Type`. 
```

Comment:
A subtle but important design choice here is that the `Type` type must grow to contain all information about any kind of type that a programmer might want to query, even if that information is not applicable to the underlying type at runtime; for example, you can access a function parameter list for all types, even if the underlying type is not a function type. I think it's important to explicitly underscore this design choice here in the opening design discussion of `Type` and justify why that's the right choice.